modules/platforms/cpp/core/src/impl/cache/cache_affinity_impl.cpp

```diff
@@ -0,0 +1,165 @@
+
+                out.Synchronize();
+
+                IgniteError err;
+                InStreamOutStream(Command::MAP_PARTITIONS_TO_NODES, *memIn.Get(), *memOut.Get(), err);
+                IgniteError::ThrowIfNeeded(err);
+
+                interop::InteropInputStream inStream(memOut.Get());
+                binary::BinaryReaderImpl reader(&inStream);
+
+                std::map<int, ClusterNode> ret;
+
+                int32_t cnt = reader.ReadInt32();
+                for (int32_t i = 0; i < cnt; i++)
+                {
+                    int key = reader.ReadInt32();
+                    ClusterNode val(GetEnvironment().GetNode(reader.ReadGuid()));
+
+                    ret.insert(std::pair<int, ClusterNode>(key, val));
+                }
+
+                return ret;
+            }
+
+            std::list<ClusterNode> CacheAffinityImpl::MapPartitionToPrimaryAndBackups(int part)
+            {
+                SharedPointer<interop::InteropMemory> memIn = GetEnvironment().AllocateMemory();
+                SharedPointer<interop::InteropMemory> memOut = GetEnvironment().AllocateMemory();
+                interop::InteropOutputStream out(memIn.Get());
+                binary::BinaryWriterImpl writer(&out, GetEnvironment().GetTypeManager());
+
+                writer.WriteObject(part);
+
+                out.Synchronize();
+
+                IgniteError err;
+                InStreamOutStream(Command::MAP_PARTITION_TO_PRIMARY_AND_BACKUPS, *memIn.Get(), *memOut.Get(), err);
+                IgniteError::ThrowIfNeeded(err);
+
+                interop::InteropInputStream inStream(memOut.Get());
+                binary::BinaryReaderImpl reader(&inStream);
+
+                std::list<ClusterNode> ret;
+                int32_t cnt = reader.ReadInt32();
+                for (int32_t i = 0; i < cnt; i++)
+                    ret.push_back(GetEnvironment().GetNode(reader.ReadGuid()));
+
+                return ret;
+            }
+
+            std::vector<int> CacheAffinityImpl::GetPartitions(int32_t opType, ClusterNode node)
+            {
+                SharedPointer<interop::InteropMemory> memIn = GetEnvironment().AllocateMemory();
+                SharedPointer<interop::InteropMemory> memOut = GetEnvironment().AllocateMemory();
+                interop::InteropOutputStream out(memIn.Get());
+                binary::BinaryWriterImpl writer(&out, GetEnvironment().GetTypeManager());
+
+                writer.WriteGuid(node.GetId());
+
+                out.Synchronize();
+
+                IgniteError err;
+                InStreamOutStream(opType, *memIn.Get(), *memOut.Get(), err);
+                IgniteError::ThrowIfNeeded(err);
+
+                interop::InteropInputStream inStream(memOut.Get());
+                binary::BinaryReaderImpl reader(&inStream);
+
+                std::vector<int> ret;
+
+                reader.ReadInt8();
+                int32_t cnt = reader.ReadInt32();
```

Comment:
Do not forget to call `reserve()` on `std::vector` once you find out the size of it.

Reply:
Fixed.